File: 2022/day07.cpp
// https://adventofcode.com/2022/day/7

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <iostream>
#include <print>
#include <ranges>

using Folder = aoc::static_vector<usize, 11>;
using Filesystem = std::pair<Vec<Folder>, Vec<u32>>;
constexpr let FILE_ID_START = usize{1} << 16;

namespace stdv = std::views;
namespace stdr = std::ranges;

auto parse(String const& filename) -> Filesystem {
  auto folder_ids = aoc::name_to_id{};
  auto file_ids = aoc::name_to_id{FILE_ID_START};
  let get_root_dir = [] {
    auto path = aoc::static_vector<String, 11>{};
    path.emplace_back("/");
    return path;
  };
  auto current_dir = get_root_dir();
  auto folders = Vec<Folder>{};
  folders.emplace_back();
  auto file_sizes = Vec<u32>{};
  auto current_dir_id = folder_ids.intern(current_dir[0]);

  for (str line : aoc::views::read_lines(filename)) {
    if (line.starts_with("$ ")) {
      let command = line.substr(2);
      if (command.starts_with("cd ")) {
        let dir = command.substr(3);
        if (dir == "/") {
          current_dir = get_root_dir();
        } else if (dir == "..") {
          current_dir.pop_back();
        } else {
          current_dir.emplace_back(dir);
        }
        current_dir_id = folder_ids.intern(aoc::ranges::join(current_dir, '/'));
      } else {
        // ls, handled on following lines
      }
    } else {
      folders.resize(folder_ids.new_size(folders.size()));
      let[first, name] = aoc::split_once<String>(line, " ");
      if (first == "dir") {
        auto subfolder = current_dir;
        subfolder.push_back(std::move(name));
        folders[current_dir_id].push_back(
            folder_ids.intern(aoc::ranges::join(subfolder, '/')));
      } else {
        auto filename = current_dir;
        filename.push_back(std::move(name));
        folders[current_dir_id].push_back(
            file_ids.intern(aoc::ranges::join(filename, '/')));

        let size = aoc::to_number<u32>(first);
        file_sizes.push_back(size);
      }
    }
  }

  return {std::move(folders), std::move(file_sizes)};
}

fn get_folder_size(Vec<Folder> const& folders, Vec<u32> const& file_sizes,
                   usize folder_id) -> u32 {
  return aoc::ranges::accumulate(
      folders[folder_id] | stdv::transform([&](usize id) {
        if (id >= FILE_ID_START) {
          return file_sizes[id - FILE_ID_START];
        } else {
          return get_folder_size(folders, file_sizes, id);
        }
      }),
      u32{});
}

fn solve_case1(Filesystem const& filesystem) -> u32 {
  let & [ folders, file_sizes ] = filesystem;
  return aoc::ranges::accumulate(
      Range{0uz, folders.size()} | //
          stdv::transform([&](usize folder_id) {
            return get_folder_size(folders, file_sizes, folder_id);
          }) |
          stdv::filter([](u32 folder_size) { return folder_size <= 100000; }),
      u32{});
}

fn solve_case2(Filesystem const& filesystem) -> u32 {
  let & [ folders, file_sizes ] = filesystem;
  let used_space = get_folder_size(folders, file_sizes, 0);
  let free_space = 70000000 - used_space;
  let min_delete_space = 30000000 - free_space;
  return stdr::min( //
      Range{1uz, folders.size()} |
      stdv::transform([&](usize folder_id) {
        return get_folder_size(folders, file_sizes, folder_id);
      }) |
      stdv::filter(
          [&](u32 folder_size) { return folder_size >= min_delete_space; }));
}

int main() {
  std::println("Part 1");
  let example = parse("day07.example");
  AOC_EXPECT_RESULT(95437, solve_case1(example));
  let input = parse("day07.input");
  AOC_EXPECT_RESULT(1453349, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(24933642, solve_case2(example));
  AOC_EXPECT_RESULT(2948823, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day07.rs
use aoc::string::NameToId;
use arrayvec::ArrayVec;

// A filesystem is a list of folders and a list of file sizes
// where each folder contains a list of IDs of files and folders.
// The two are differentiated by file IDs starting at a higher number.
type Folder = ArrayVec<usize, 11>;
type Filesystem = (Vec<Folder>, Vec<u32>);
const FILE_ID_START: usize = 1 << 16;

fn parse(filename: &str) -> Filesystem {
    let mut folder_ids = NameToId::new();
    let mut file_ids = NameToId::from(FILE_ID_START);
    let get_root_dir = || {
        let mut path = ArrayVec::<String, 11>::new();
        path.push("/".to_string());
        path
    };
    let mut current_dir = get_root_dir();
    let mut folders = Vec::new();
    folders.push(Folder::default());
    let mut file_sizes = Vec::new();
    let mut current_dir_id = folder_ids.intern(&current_dir[0]);

    for line in aoc::file::read_lines(filename) {
        if let Some((_, command)) = line.split_once("$ ") {
            if let Some((_, dir)) = command.split_once("cd ") {
                current_dir = match dir {
                    "/" => get_root_dir(),
                    ".." => {
                        current_dir.pop();
                        current_dir
                    }
                    dir => {
                        current_dir.push(dir.to_string());
                        current_dir
                    }
                };
                current_dir_id = folder_ids.intern(&current_dir.join("/"));
            } else {
                // ls, handled on following lines
            }
        } else {
            folders.resize(folder_ids.new_len(folders.len()), Folder::default());
            let (first, name) = line.split_once(" ").unwrap();
            let name = name.to_string();
            folders[current_dir_id].push(match first {
                "dir" => {
                    let mut subfolder = current_dir.clone();
                    subfolder.push(name);
                    folder_ids.intern(&subfolder.join("/"))
                }
                size => {
                    let size = size.parse::<u32>().unwrap();
                    let mut filename = current_dir.clone();
                    filename.push(name);
                    let id = file_ids.intern(&filename.join("/"));
                    file_sizes.push(size);
                    id
                }
            });
        }
    }

    return (folders, file_sizes);
}

fn get_folder_size(folders: &Vec<Folder>, file_sizes: &Vec<u32>, folder_id: usize) -> u32 {
    folders[folder_id]
        .iter()
        .map(|&id| {
            if (id >= FILE_ID_START) {
                file_sizes[id - FILE_ID_START]
            } else {
                get_folder_size(folders, file_sizes, id)
            }
        })
        .sum::<u32>()
}

fn solve_case1((folders, file_sizes): &Filesystem) -> u32 {
    (0..(folders.len()))
        .map(|folder_id| get_folder_size(folders, file_sizes, folder_id))
        .filter(|&folder_size| folder_size <= 100000)
        .sum()
}

fn solve_case2((folders, file_sizes): &Filesystem) -> u32 {
    let used_space = get_folder_size(folders, file_sizes, 0);
    let free_space = 70000000 - used_space;
    let min_delete_space = 30000000 - free_space;

    (1..(folders.len()))
        .map(|folder_id| get_folder_size(folders, file_sizes, folder_id))
        .filter(|&folder_size| folder_size >= min_delete_space)
        .min()
        .unwrap()
}

fn main() {
    println!("Part 1");
    let example = parse("day07.example");
    assert_eq!(95437, solve_case1(&example));
    let input = parse("day07.input");
    assert_eq!(1453349, solve_case1(&input));

    println!("Part 2");
    assert_eq!(24933642, solve_case2(&example));
    assert_eq!(2948823, solve_case2(&input));
}
